File: solutions/Golang/add_binary_bit_operations.go
package main

import (
	"fmt"
	"strconv"
)

/*
Bit operations approach - convert to uint64, add, convert back to binary.
Time: O(max(len(a), len(b)))
Space: O(max(len(a), len(b))) for result
*/
func addBinaryBitOperations(a string, b string) string {
	numA, _ := strconv.ParseUint(a, 2, 64)
	numB, _ := strconv.ParseUint(b, 2, 64)
	total := numA + numB
	return strconv.FormatUint(total, 2)
}

func main() {
	fmt.Println(addBinaryBitOperations("11", "1"))      // "100"
	fmt.Println(addBinaryBitOperations("1010", "1011")) // "10101"
	fmt.Println(addBinaryBitOperations("0", "0"))       // "0"
}

File: solutions/Java/AddBinaryBitOperations.java
public class AddBinaryBitOperations {
    /*
    Bit operations approach - convert to long, add, convert back to binary.
    Time: O(max(a.length(), b.length()))
    Space: O(max(a.length(), b.length())) for result
    */
    public static String addBinaryBitOperations(String a, String b) {
        long numA = Long.parseLong(a, 2);  // Convert binary string to long
        long numB = Long.parseLong(b, 2);
        long total = numA + numB;
        return Long.toBinaryString(total);
    }

    public static void main(String[] args) {
        System.out.println(addBinaryBitOperations("11", "1"));      // "100"
        System.out.println(addBinaryBitOperations("1010", "1011")); // "10101"
        System.out.println(addBinaryBitOperations("0", "0"));       // "0"
    }
}
